src/main/java/hudson/plugins/timestamper/accessor/TimestampLogFileLineAccessor.java

```diff
@@ -0,0 +1,67 @@
+package hudson.plugins.timestamper.accessor;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import hudson.plugins.timestamper.Timestamp;
+import hudson.plugins.timestamper.action.TimestampsActionOutput;
+import hudson.plugins.timestamper.io.LogFileReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Optional;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Abstraction for retrieving timestamps and log file lines from completed builds. Timestamp records
+ * can be stored in different files and formats for different types of builds. Consumers that wish
+ * to retrieve such records should use this interface rather than directly opening the corresponding
+ * log file.
+ */
+@Restricted(NoExternalUse.class)
+public interface TimestampLogFileLineAccessor extends Closeable {
+
+    /**
+     * Value class representing a single record in the abstract, regardless of how the storage of
+     * that record is implemented.
+     */
+    class TimestampLogFileLine {
+        private final Optional<Timestamp> timestamp;
+        private final Optional<String> logFileLine;
+
+        public TimestampLogFileLine(Optional<Timestamp> timestamp, Optional<String> logFileLine) {
+            this.timestamp = checkNotNull(timestamp);
+            this.logFileLine = checkNotNull(logFileLine);
+        }
+
+        /** Return the timestamp associated with the record, if present. */
+        public Optional<Timestamp> getTimestamp() {
+            return timestamp;
+        }
+
+        /**
+         * Return the log file line associated with the record, if present. This log file line may
+         * contain console notes. It is the caller's responsibility to remove any console notes if
+         * desired.
+         */
+        public Optional<String> getLogFileLine() {
+            return logFileLine;
+        }
+    }
+
+    /**
+     * Retrieve the {@link LogFileReader} associated with this accessor, likely for use with {@link
+     * TimestampsActionOutput.LineCountSupplier}.
+     */
+    LogFileReader getLogFileReader();
+
+    /** Skip forward one line in the associated record file(s). */
+    void skipLine() throws IOException;
+
+    /**
+     * Retrieve a log file line and its associated timestamp. While typically both a timestamp and a
+     * log file line will be present, this API is resilient to edge cases in which one or the other
+     * is not present. When neither a timestamp nor a log file line are present, EOF has been
+     * reached and callers should stop retrieving further records.
```

Comment:
`EOFException`? Might be wise to document that if so.

Reply:
I thought about that and other improvements, but I didn't want to change the existing code for Freestyle jobs out of fear of introducing a regression. So I made the interface reflect the existing code rather than designing the ideal interface. Perhaps I'll look at improving the existing code (and therefore the interface) as part of some future code cleanup change. The good news is that the interface is restricted to no external use, so we can evolve it without worrying about breaking any other plugins.

Reply:
I just meant to clarify the Javadoc. If “one or the other is not present”, what exactly happens? How do you know that "neither a timestamp nor a log file are present” or “EOF has been reached”?

Reply:
These are all good questions which reflect the confusing design and implementation of the existing code.

> If “one or the other is not present”, what exactly happens?

If only one of the two is present, then [the output will just be missing half of the requested information](https://github.com/jenkinsci/timestamper-plugin/blob/38f17c77247055af3663e8cc12f74878acedaf49/src/main/java/hudson/plugins/timestamper/action/TimestampsActionOutput.java#L173-L186) (!). I guess I can update the Javadoc to say "In such cases, consumers have a choice as to whether to discard the record or return incomplete information to the user", not that I particularly understand how this situation could occur in the first place. Does that sound OK?

>  How do you know that "neither a timestamp nor a log file are present” or “EOF has been reached”?

The existing `readNextLine` method [returns `Optional.empty()` when `!timestamp.isPresent() && !logFileLine.isPresent()`](https://github.com/jenkinsci/timestamper-plugin/blob/38f17c77247055af3663e8cc12f74878acedaf49/src/main/java/hudson/plugins/timestamper/action/TimestampsActionOutput.java#L188-L191), which [signals its caller to `break` and stop reading new lines](https://github.com/jenkinsci/timestamper-plugin/blob/38f17c77247055af3663e8cc12f74878acedaf49/src/main/java/hudson/plugins/timestamper/action/TimestampsActionOutput.java#L127-L130). I don't think this existing code is ideal, but I tried to modify it as little as possible to add Pipeline support. I actually think the existing Javadoc explains the existing (ugly) code pretty accurately, unless you have any further suggestions.

Reply:
> Does that sound OK?

Whatever you think best. Just thought I would point to a place where the existing description sounded too vague.

Reply:
Sure, I'll go ahead and add that extra sentence. Perhaps I was subconsciously being vague on purpose to avoid having to discuss the confusing nature of the existing code… In any case, I am just being extremely conservative trying to minimize code change because I am a new maintainer and don't want to introduce regressions. Once I become more familiar and confident with this plugin, I'll be emboldened to make more invasive cleanup changes that can result in better abstractions.